token/program2/src/instruction.rs

```diff
@@ -196,14 +220,58 @@ impl TokenInstruction {
         }
         Ok(match input[0] {
             0 => {
-                if input.len() < size_of::<u8>() + size_of::<u64>() + size_of::<u8>() {
+                if input.len()
+                    < size_of::<u8>() + size_of::<u64>() + size_of::<u8>() + size_of::<bool>()
+                {
                     return Err(TokenError::InvalidInstruction.into());
                 }
+                let mut input_len = 0;
+                input_len += size_of::<u8>();
+
                 #[allow(clippy::cast_ptr_alignment)]
-                let amount = unsafe { *(&input[size_of::<u8>()] as *const u8 as *const u64) };
-                let decimals =
-                    unsafe { *(&input[size_of::<u8>() + size_of::<u64>()] as *const u8) };
-                Self::InitializeMint { amount, decimals }
+                let amount = unsafe { *(&input[input_len] as *const u8 as *const u64) };
+                input_len += size_of::<u64>();
+                let decimals = unsafe { *(&input[input_len] as *const u8) };
+                input_len += size_of::<u8>();
+
+                let mint_authority = match input[input_len] {
```

Comment:
A single inline function for unpacking/packing a `COption<Pubkey>` would be nice.   This unpack/pack code is hard to read *and* unsafe.    The more we can reduce it the better

Reply:
The more complex the token program gets the more unsavory this stuff gets

Reply:
Attempt: https://github.com/solana-labs/solana-program-library/pull/311

Reply:
Done